Std/Data/BitVec/Lemmas.lean

```diff
@@ -205,11 +237,35 @@ protected theorem extractLsb_ofNat (x n : Nat) (hi lo : Nat) :
   rw [← testBit_toNat, getLsb, getLsb]
   simp
 
-/-! ### not-/
+/-! ### not -/
+
+theorem not_def {x : BitVec v} : ~~~x = allOnes v ^^^ x := rfl
 
-@[simp] proof_wanted toNat_not {x : BitVec v} : (~~~x).toNat = 2^v - x.toNat
+@[simp] theorem toNat_not {x : BitVec v} : (~~~x).toNat = 2^v - (x.toNat + 1) := by
+  rw [not_def, toNat_xor]
+  apply Nat.eq_of_testBit_eq
+  intro i
+  simp only [toNat_allOnes, Nat.testBit_xor, Nat.testBit_two_pow_sub_one]
+  match h : BitVec.toNat x with
+  | 0 => simp
+  | y+1 =>
+    rw [Nat.succ_eq_add_one] at h
+    rw [← h]
+    rw [Nat.testBit_two_pow_sub_succ (toNat_lt _)]
+    · cases w : decide (i < v)
+      · simp at w
+        simp [w]
+        rw [Nat.testBit_lt_two_pow]
+        calc BitVec.toNat x < 2 ^ v := toNat_lt _
+          _ ≤ 2 ^ i := Nat.pow_le_pow_of_le_right Nat.zero_lt_two w
+      · simp
```

Comment:
I'm glad you corrected the statement; shows the danger of proof_wanted.

I'd be tempted to write the right hand side as `(~~~x).toNat = (2^v - 1) - x.toNat` just because there are `Nat` theorems abort `2^n - 1` since it's `n` ones in binary.

Reply:
Changed the statement (but only minimally the proof).